File: Projet/projet.js
document.addEventListener("DOMContentLoaded", function () {

    // Ajout d'un gestionnaire d'événements sur l'icône de l'acceuil
    const accueilLink = document.getElementById('accueil-link');
    accueilLink.addEventListener('click', function(event) {
        event.preventDefault(); // Pour empêcher le comportement par défaut du lien
        displayTweet(); // Appeler la fonction displayTweets lorsque l'icône de la maison est cliquée
    });
    
    // on récupère le pseudo depuis localStorage
    const pseudo = localStorage.getItem('pseudo');

    // Remplir la section du profil avec le pseudo
    const profileSection = document.getElementById('profileSection');
    profileSection.innerHTML = `
        <p>Pseudo: ${pseudo}</p>
        <div class="profile-picture">
            <img src="/projet/img/${pseudo}.webp" alt="Photo de profil">
        </div>
    `;


    displayTweet(pseudo)
    displayHashtags()
});


function displayTweet(pseudo){
   
   fetch('http://127.0.0.1:5000/tweets')
   .then(response => response.json())
   .then(data => {
       if (data.success) {
           const tweets = data.tweets;
           const tweetSection = document.querySelector('.tweets');
           tweetSection.innerHTML = ''; // on efface les tweets existants

           tweets.forEach(tweet => {
               const tweetDiv = document.createElement('div');
               tweetDiv.classList.add('tweet');

                            


               const profilePicDiv = document.createElement('div');
               profilePicDiv.classList.add('profile-pic'); // pour la pdp


                const profilePicImg = document.createElement('img');
                // on définit l'attribut src de l'image en utilisant le chemin vers l'image
                profilePicImg.src = `/projet/img/${tweet.username}.webp`;
                // on definit l'attribut alt de l'image 
                profilePicImg.alt = `Photo de profil de ${tweet.username}`;
                // on ajoute la classe pour le style CSS
                profilePicImg.classList.add('profile-pic-img');

                // Ajouter l'image à la division de la photo de profil
                profilePicDiv.appendChild(profilePicImg);


                
               
                

               const tweetContentDiv = document.createElement('div');
               tweetContentDiv.classList.add('tweet-content');

               const usernameHeading = document.createElement('h3');
               usernameHeading.classList.add('username');
               usernameHeading.textContent = tweet.username;

               const tweetTextParagraph = document.createElement('p');
               tweetTextParagraph.classList.add('tweet-text');
               tweetTextParagraph.textContent = tweet.tweet_text;



               

               // partie retweet

                // on précise si c'est un retweet 
                if (tweet.retweeter) {
                    const retweetInfo = document.createElement('p');
                    retweetInfo.textContent = `Retweeté par: @${tweet.retweeter}`;
                    retweetInfo.classList.add('retweet-info'); // Ajoutez la classe pour le style CSS
                    tweetContentDiv.appendChild(retweetInfo); // Ajoutez retweetInfo à tweetContentDiv
                }



               const retweetButton = document.createElement('button');
                    retweetButton.innerHTML = '<i class="fas fa-retweet"></i> Retweet';
                    retweetButton.classList.add('retweet-button');



                    //on verifie si le boutton est deja cliqué
                    const isRetweeted = localStorage.getItem(`retweet-${tweet.tweet_text}`);
                    
                    if (isRetweeted) {
                        
                        
                        retweetButton.classList.add('retweeted');// on marque le boutton
                    }

                    retweetButton.addEventListener('click', function() {
                        retweetTweet(tweet.username,pseudo, tweet.tweet_text,retweetButton);
                        console.log('Retweet clicked!');
                    });

               tweetContentDiv.appendChild(usernameHeading);
               tweetContentDiv.appendChild(tweetTextParagraph);
               tweetContentDiv.appendChild(retweetButton);

               
               tweetDiv.appendChild(profilePicDiv);
               tweetDiv.appendChild(tweetContentDiv);

           
               tweetSection.appendChild(tweetDiv); 

               
               
           });
       }
   })
   .catch(error => {
       console.error('Error fetching tweets:', error);
   });
}


function sendTweet() {
    const tweetText = document.getElementById('tweetText').value;


    // on Vérifie si le champ de texte n'est pas vide
    if (!tweetText.trim()) {
        alert('Veuillez entrer votre tweet avant de l\'envoyer.');
        return; // on Arrête l'exécution de la fonction si le champ est vide
    }

    // Récupérer le pseudo depuis localStorage
    const pseudo = localStorage.getItem('pseudo');

    // Envoi du tweet au serveur Flask
    fetch('http://127.0.0.1:5000/tweet', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            username: pseudo,
            tweet_text: tweetText,
        }),
    })
    .then(response => response.json())
    .then(data => {

        console.log(data.tweet);
        if (data.message== "Tweet enregistré avec succès") {

            const tweetSection = document.querySelector('.tweets');

            const tweetDiv = document.createElement('div');
            tweetDiv.classList.add('tweet');

            const profilePicDiv = document.createElement('div');
            profilePicDiv.classList.add('profile-pic');
            
            
            profilePicDiv.style.backgroundColor = '#ccc'; 

            const tweetContentDiv = document.createElement('div');
            tweetContentDiv.classList.add('tweet-content');

            const usernameHeading = document.createElement('h3');
            usernameHeading.classList.add('username');
            usernameHeading.textContent = pseudo;

            const tweetTextParagraph = document.createElement('p');
            tweetTextParagraph.classList.add('tweet-text');
            tweetTextParagraph.textContent = tweetText;

            tweetContentDiv.appendChild(usernameHeading);
            tweetContentDiv.appendChild(tweetTextParagraph);

            tweetDiv.appendChild(profilePicDiv);
            tweetDiv.appendChild(tweetContentDiv);
            

            tweetSection.prepend(tweetDiv); 
        
        }

    })
    .catch(error => {
        console.error('Erreur lors de l\'envoi du tweet:', error);
    });

    // Réinitialisation du formulaire
    document.getElementById('tweetForm').reset();
}

function logout() {
    // Clear localStorage (assuming you stored the user information there)
    localStorage.clear();

    // Redirect to index.html
    window.location.href = 'index.html';
}




function retweetTweet(username, Retweeter, tweet,retweetButton) {
 // on vérifie si le boutton n'a pas deja été rt
 const isRetweeted = localStorage.getItem(`retweet-${tweet}`);
 if (!isRetweeted) {
    fetch('http://127.0.0.1:5000/retweets', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            username: username,
            Retweeter: Retweeter,
            tweet_text: tweet,
        }),
    })
    .then(response => response.json())
    .then(data => {
        console.log(data.message);
        retweetButton.classList.add('retweeted');// on modifie la couleur avec le css
        localStorage.setItem(`retweet-${tweet}`, true); // on stock le status du tweet
    })
    .catch(error => {
        console.error('Error retweeting tweet:', error);
    });
}
}

function displayHashtags() {
    fetch('http://127.0.0.1:5000/hashtags')
        .then(response => response.json())
        .then(data => {
            if (data.success) {
                const hashtags = data.hashtags;
                const trendingSection = document.querySelector('.hashtag-section');
                const uniqueHashtags = new Set(hashtags); // Utilise un ensemble sans doublon pour stocker des hashtags uniques

                trendingSection.innerHTML = ''; // Effacer le contenu existant

                uniqueHashtags.forEach(hashtag => {
                    const hashtagElement = document.createElement('div');
                    hashtagElement.classList.add('hashtag');
                    hashtagElement.textContent = hashtag;
                    trendingSection.appendChild(hashtagElement);


                    // Ajoute un événement de clic à chaque hashtag
                     hashtagElement.addEventListener('click', function() {
                        
                          displayByHashtags(hashtag); // Appele de la fonction displayByHashtags avec le hashtag cliqué
                     });
                });
            }
        })
        .catch(error => {
            console.error('Error fetching hashtags:', error);
        });
}


function displayByHashtags(hashtag) {
    
    fetch(`http://127.0.0.1:5000/displayTweetsByHashtag/${hashtag}`)
        .then(response => response.json())
        .then(data => {
            if (data.success) {
                
            const tweets = data.tweets;
            const tweetSection = document.querySelector('.tweets');
            tweetSection.innerHTML = ''; // on efface les tweets existants
            console.log(tweets);
            tweets.forEach(tweet => {
               const tweetDiv = document.createElement('div');
               tweetDiv.classList.add('tweet');

               const profilePicDiv = document.createElement('div');
               profilePicDiv.classList.add('profile-pic'); // pour la pdp


                const profilePicImg = document.createElement('img');
                // on définit l'attribut src de l'image en utilisant le chemin vers l'image
                profilePicImg.src = `/projet/img/${tweet.username}.webp`;
                // on definit l'attribut alt de l'image 
                profilePicImg.alt = `Photo de profil de ${tweet.username}`;
                // on ajoute la classe pour le style CSS
                profilePicImg.classList.add('profile-pic-img');

                // Ajouter l'image à la division de la photo de profil
                profilePicDiv.appendChild(profilePicImg);
                

               const tweetContentDiv = document.createElement('div');
               tweetContentDiv.classList.add('tweet-content');

               const usernameHeading = document.createElement('h3');
               usernameHeading.classList.add('username');
               usernameHeading.textContent = tweet.username;

               const tweetTextParagraph = document.createElement('p');
               tweetTextParagraph.classList.add('tweet-text');
               tweetTextParagraph.textContent = tweet.tweet_text;
     
               tweetContentDiv.appendChild(usernameHeading);
               tweetContentDiv.appendChild(tweetTextParagraph);
               
               tweetDiv.appendChild(profilePicDiv);
               tweetDiv.appendChild(tweetContentDiv);

           
               tweetSection.appendChild(tweetDiv); 
                });
            }
        })
        .catch(error => {
            console.error('Error fetching hashtags:', error);
        });
}

File: Projet/projet.py
from flask import Flask, request, jsonify
from flask_cors import CORS
import redis

app = Flask(__name__)
CORS(app)
redis_client = redis.StrictRedis(host='localhost', port=6379, db=0)

@app.route('/register', methods=['POST'])
def register_user():
    data = request.json  #on s'assure d'envoyer les données en tant que JSON depuis votre front-end
    username = data.get('username')
    password = data.get('password')

    # Vérifie si le nom d'utilisateur est déjà pris
    if redis_client.sismember('usernames', username):
        return jsonify({"message": "Nom d'utilisateur déjà pris. Veuillez en choisir un autre."}), 400

    # Enregistre les informations de l'utilisateur dans Redis
    
    user_key = f'id-{username}'
    redis_client.hmset(user_key, {'username': username, 'password': password})
    redis_client.sadd('usernames', username)

    return jsonify({"message": "Utilisateur enregistré avec succès"}), 201

@app.route('/login', methods=['POST'])
def login_user():
    data = request.json  #on s'assure d'envoyer les données en tant que JSON depuis votre front-end
    username = data.get('username')
    password = data.get('password')

    user_key = f'id-{username}'

    # Vérifiez si le nom d'utilisateur existe dans Redis
    if not redis_client.exists(user_key):
        return jsonify({"success": False, "message": "Nom d'utilisateur incorrect"}), 401

    # Récupérez les informations de l'utilisateur depuis Redis
    user_info = redis_client.hgetall(user_key)

    # Vérifiez si le mot de passe correspond
    if password == user_info[b'password'].decode():
        return jsonify({"success": True, "message": "Connexion réussie"}), 200
    else:
        return jsonify({"success": False, "message": "Mot de passe incorrect"}), 401
    
@app.route('/tweet', methods=['POST'])
def tweet():
    try:
        data = request.get_json()
        username = data.get('username')
        tweet_text = data.get('tweet_text')


        # on extrait le hashtag
        hashtags = [word for word in tweet_text.split() if word.startswith('#')]
        # on verifie si il y a des hashtags
        if hashtags:
            # Store each hashtag in Redis
            for hashtag in hashtags:
                hashtag_key = redis_client.incr('hashtag_id_counter')
                hashtag_key = f'hashtag-{hashtag_key}'
                redis_client.set(hashtag_key, hashtag)
        
        id_tweet = redis_client.incr('tweet_id_counter')
        tweet_key = f'tweet-{id_tweet}'

        redis_client.hmset(tweet_key, {'username': username, 'tweet': tweet_text})
        redis_client.sadd(username, tweet_key)

        return jsonify({"success": True, "message": "Tweet enregistré avec succès","hashtag": hashtag}), 201
    except Exception as e:
        print(f"Erreur lors de la gestion de la requête : {str(e)}")
        return jsonify({"success": False, "message": "Erreur lors de la gestion de la requête"}), 500


@app.route('/tweets', methods=['GET'])
def get_tweets():
    try:
        # on récupère les clés des tweets et retweets
        tweet_keys = redis_client.keys('tweet-*')

        # on crée un tableau de tweets
        tweets = []

        # on boucle et on remplit le tableau avec les tweets et les retweets
        for tweet_key in tweet_keys:
            tweet_data = redis_client.hgetall(tweet_key)
            tweet = {
                'username': tweet_data[b'username'].decode(),
                'tweet_text': tweet_data[b'tweet'].decode()
            }
            # Si le tweet a un champ 'retweeter' alors c'est un retweet donc on l'add
            if b'retweeter' in tweet_data:
                tweet['retweeter'] = tweet_data[b'retweeter'].decode()
            tweets.append(tweet)

        return jsonify({"success": True, "tweets": tweets}), 200
    except Exception as e:
        print(f"Error fetching tweets: {str(e)}")
        return jsonify({"success": False, "message": "Erreur d'affichage"}), 500


@app.route('/tweets/<username>', methods=['GET'])
def get_user_tweets(username):
    try:
        # on veifie si l'user existe
        if not redis_client.exists(f'id-{username}'):
            return jsonify({"success": False, "message": "utilisateur innexistant"}), 404

        # on reccup tous ses tweets
        tweet_keys = redis_client.smembers(username)
        tweets = []

        # onremplit le tableau de tweets et on le renvoie au compte.js
        for tweet_key in tweet_keys:
            tweet_data = redis_client.hgetall(tweet_key)
            tweet = {
                'username': tweet_data[b'username'].decode(),
                'tweet_text': tweet_data[b'tweet'].decode()
            }
            tweets.append(tweet)

        return jsonify({"success": True, "tweets": tweets}), 200
    except Exception as e:
        print(f"Error fetching user tweets: {str(e)}")
        return jsonify({"success": False, "message": "Error d'affichage"}), 500


@app.route('/retweets', methods=['POST'])
def retweet():
    try:
        data = request.get_json()
        username = data.get('username')
        retweeter = data.get('Retweeter')
        tweet_text = data.get('tweet_text')
        
        id_tweet = redis_client.incr('tweet_id_counter')
        tweet_key = f'tweet-{id_tweet}'

        # on crée le retweet de manière à ce qu'il soit traité 
        redis_client.hmset(tweet_key, {'username': username, 'tweet': tweet_text, 'retweeter': retweeter})
        redis_client.sadd(retweeter, tweet_key)

        return jsonify({"success": True, "message": "reTweet enregistré avec succès"}), 201
    except Exception as e:
        print(f"Erreur lors de la gestion de la requête : {str(e)}")
        return jsonify({"success": False, "message": "Erreur lors de la gestion de la requête"}), 500


    
@app.route('/hashtags', methods=['GET'])
def get_hashtags():
    try:
        # Récupérer tous les hashtags stockés dans Redis
        hashtag_keys = redis_client.keys('hashtag-*')

        # Créer un ensemble pour stocker des hashtags uniques
        hashtags = set()

        for hashtag_key in hashtag_keys:
            hashtag = redis_client.get(hashtag_key).decode()
            hashtags.add(hashtag)

        return jsonify({"success": True, "hashtags": list(hashtags)}), 200
    except Exception as e:
        print(f"Error fetching hashtags: {str(e)}")
        return jsonify({"success": False, "message": "Erreur lors de la récupération des hashtags"}), 500

if __name__ == '__main__':
    app.run(debug=True)
